src/daemon/scheduler.rs

```diff
@@ -169,230 +126,161 @@ fn make_cas_event_triggers(event_queue: Arc<MessageQueue>, ca_manager: Arc<CaMan
-            requeue_time_test()
-        } else {
-            requeue_time()
-        };
-
-        error!(
-            "Failed to synchronize CA '{}' with its parent '{}', error: {}",
-            ca, parent, e
-        );
-        event_queue.reschedule_sync_parent(ca, parent, requeue_time);
-    }
-}
-
-fn make_cas_republish(ca_server: Arc<CaManager>, event_queue: Arc<MessageQueue>) -> ScheduleHandle {
-    SkippingScheduler::run(
-        SCHEDULER_INTERVAL_SECONDS_REPUBLISH,
-        "CA certificate republish",
-        move || {
-            let rt = Runtime::new().unwrap();
-            rt.block_on(async {
-                debug!("Triggering background republication for all CAs, note this may be a no-op");
-                match ca_server.republish_all().await {
-                    Err(e) => error!("Background republishing of MFT and CRLs failed: {}", e),
-                    Ok(cas) => {
-                        for ca in cas {
-                            info!("Re-issued MFT and CRL for CA: {}", ca);
-                            event_queue.schedule_sync_repo(ca);
-                        }
-                    }
-                }
-            })
-        },
-    )
-}
-
-fn make_cas_objects_renew(ca_server: Arc<CaManager>, actor: Actor) -> ScheduleHandle {
-    SkippingScheduler::run(SCHEDULER_INTERVAL_SECONDS_ROA_RENEW, "CA ROA renewal", move || {
-        let rt = Runtime::new().unwrap();
-        rt.block_on(async {
-            debug!(
-                "Triggering background renewal for about to expire objects issued by all CAs, note this may be a no-op"
-            );
-            if let Err(e) = ca_server.renew_objects_all(&actor).await {
-                error!("Background re-issuing of about to expire objects failed: {}", e);
+    /// Queues tasks for background jobs when the server is started
+    async fn queue_start_tasks(&self) -> KrillResult<()> {
+        // If there are only a few CAs in this krill instance, then we
+        // will just want to re-sync them with their parents and repository
+        // on start up.
+        //
+        // If there are many, then will apply some random delays (jitter)
+        // to avoid a thundering herd. Note that the operator can always
+        // choose to run bulk operations manually if they know that they
+        // cannot wait.
+        let ca_list = self.ca_manager.ca_list(&self.system_actor)?;
+        let cas = ca_list.cas();
+
+        debug!("Adding tasks at start up");
+
+        let mut use_jitter = cas.len() >= SCHEDULER_USE_JITTER_CAS_THRESHOLD;
+
+        for summary in cas {
+            let ca = self.ca_manager.get_ca(summary.handle()).await?;
+
+            use_jitter = use_jitter || ca.nr_parents() >= self.config.ca_refresh_parents_batch_size;
+
+            debug!("Adding tasks for CA {}, using jitter: {}", ca.handle(), use_jitter);
+
+            // Plan a regular sync for each parent. Spread these out if there
```

Comment:
Technically the won't necessarily be spread (i.e. evenly) out in the range, they will randomly (which could be uneven, or even clumped) adjusted within the range. If the desire is to spread them evenly then the logic should be changed.

Reply:
There is no need for this to be an even spread, it's just an easy way to ensure that the syncs are not all scheduled for the same moment.

This is even more important for the case where we have too many local CAs - because -likely- they will have local parents as well. E.g. 1000+ hosted CAs.

The semantics could be made smarter, but then it would need a lot more work to analyse all CAs and which parents they actually have, and how long ago the last sync (before restart) was.. and I don't think the complication is worth it.

Now - in case of a few CAs and parents per CA we just do a sync at start up, and otherwise we add some random delays. Using the 12 hour default and the mostly flat rather than gaussian random curve that I assume the random number generator will use this should spread things out more than enough.

Normally these syncs are not that time critical. And the user can still force rescheduling of the syncs using `krillc bulk refresh` if they have a reason to assume that there is a need.